src/charm.py

```diff
@@ -253,6 +228,31 @@ def sha256(hashable: Any) -> str:
             if lokitool_output.stderr:
                 logger.error(f"lokitool (err): {lokitool_output.stderr.read().strip()}")
 
+    def _update_datasource_exchange(self) -> None:
+        """Update the grafana-datasource-exchange relations."""
+        if not self.unit.is_leader():
+            return
+
+        # we might have multiple grafana-source relations, this method collects them all and returns a mapping from
+        # the `grafana_uid` to the contents of the `datasource_uids` field
+        # for simplicity, we assume that we're sending the same data to different grafanas.
+        # read more in https://discourse.charmhub.io/t/tempo-ha-docs-correlating-traces-metrics-logs/16116
+        grafana_uids_to_units_to_uids = self.grafana_source.get_source_uids()
+        raw_datasources: List[DatasourceDict] = []
+
+        for grafana_uid, ds_uids in grafana_uids_to_units_to_uids.items():
+            for _, ds_uid in ds_uids.items():
+                raw_datasources.append({"type": "loki", "uid": ds_uid, "grafana_uid": grafana_uid})
+        self.coordinator.datasource_exchange.publish(datasources=raw_datasources)
+
+    def _reconcile(self):
+        # This method contains unconditional update logic, i.e. logic that should be executed
+        # regardless of the event we are processing.
+        if self._nginx_container.can_connect():
+            self._set_alerts()
+        self._ensure_lokitool()
+        self._update_datasource_exchange()
```

Comment:
There is some information we've lost with this re-factor, for instance:

```python
logger.info("Ingress for app ready on '%s'", event.url)
```

Despite of the fact I like this pattern because makes everything clean, we need to come up with a standardised way to process information which is only available in the specific event.

Reply:
see https://github.com/canonical/mimir-coordinator-k8s-operator/pull/100#discussion_r1883845399